### ISSUE TITLE
Why is my train_loss nan

### ISSUE THREAD

**Issue by yangershuai627** [ultralytics/ultralytics#21020]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

When using the custom module A+B alone, the train_loss is not nan. When using the custom module C alone, the train_loss is not nan either. When using module A+B+C, the train_loss is nan, as shown in the following figure.
![Image](https://github.com/user-attachments/assets/c8e4803a-48d7-4cba-a832-27c197c689d9)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yangershuai627, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

_This is an automated response. An Ultralytics engineer will assist you here soon!_

**Comment by yangershuai627** [comment]
@Y-T-G Do you know how to solve this problem?

**Comment by Y-T-G** [comment]
You can try reducing learning rate

**Comment by yangershuai627** [comment]
@Y-T-G Learning rate has dropped to a very low level, but train_loss is still nan.
```
from ultralytics import YOLO

# Load a model
model = YOLO("custom_model.yaml").load("yolov8n-obb.pt")

# Train the model
results = model.train(data="dataset.yaml", 
                      epochs=300,
                      batch=16,
                      device=[0,2,4,9],
                      imgsz=800,
                      val=False,
                      lr0=0.0000000001
                      )
```

![Image](https://github.com/user-attachments/assets/e4ab8562-8aa8-4627-9be1-e36d18ac1d47)

**Comment by Y-T-G** [comment]
Without setting optimizer, learning rate doesn't take effect

**Comment by yangershuai627** [comment]
The optimizer has been set, but the learning rate remains nan
![Image](https://github.com/user-attachments/assets/6a9db4ed-315e-4262-8272-00aa35319567)

![Image](https://github.com/user-attachments/assets/09de95aa-e576-48c5-afb7-ae5449a13576)

**Comment by glenn-jocher** [comment]
Your learning rate of `0.0000000001` is extremely low and may cause numerical instability. Try using the default learning rates instead - remove the `lr0` parameter and let the system auto-select, or use `lr0=0.01` for SGD or `lr0=0.001` for Adam. Also, when combining custom modules that work individually, check for gradient explosion/vanishing - you might need gradient clipping with `gradient_clip_val=10.0` in your training arguments to prevent NaN losses.

**Comment by yangershuai627** [comment]
Hello @glenn-jocher , there is no parameter named gradient_clip_val in default.yaml of YOLOv8.

**Comment by Y-T-G** [comment]
You can try training with `amp=False`

**Comment by yangershuai627** [comment]
`amp=False` has been set and doesn't work.

**Comment by glenn-jocher** [comment]
You're correct that `gradient_clip_val` isn't a standard YOLO parameter. Since your individual modules work fine but combining them causes NaN losses, this suggests architectural compatibility issues. Try using default learning rates first (`lr0=0.01` for SGD), then debug your custom modules by adding print statements to check for extreme values in the forward pass outputs - NaN often occurs when modules produce infinite or very large intermediate values that compound when combined.

**Comment by yangershuai627** [comment]
I found again that running the model train_loss with 1080Ti is normal, but running it with two 1080Ti together will result in nan.

**Comment by glenn-jocher** [comment]
This is a classic multi-GPU distributed training issue. The 1080Ti GPUs may have compatibility problems with distributed training - try using `device=0` (single GPU) first to confirm it works, then if you need multi-GPU, ensure all GPUs are identical models with the same CUDA compute capability and consider updating your CUDA drivers, as older GPUs like 1080Ti can have synchronization issues in distributed setups.
